### PR TITLE
Further filter detected slow broker against pre-defined flush time threshold.

### PR DESCRIPTION
**Pull request by kidkun**
Addresses the issue #1185 .

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java [hunk 1]
```diff
@@ -174,6 +184,7 @@ public SlowBrokerFinder() {
     }
 
     Set<BrokerEntity> detectMetricAnomalies = getMetricAnomalies(historicalLogFlushTimeMetricValues, currentLogFlushTimeMetricValues);
+    detectMetricAnomalies.retainAll(detectMetricAnomaliesFromThreshold(currentLogFlushTimeMetricValues, _logFlushTimeThresholdMs));
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java:187]: Can we avoid passing private variable `_logFlushTimeThresholdMs` to non-static function `detectMetricAnomaliesFromThreshold` and update the name of the function to indicate its custom role regarding log flush time?

Actually it is a little confusing that we have another function in this class called `getMetricAnomalies`, which is also used for detecting metric anomalies -- can we merge/rename them to make the logic more clear?

Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java:187]: Nit: `detectMetricAnomalies` -> `detectedMetricAnomalies`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java [hunk 2]
```diff
@@ -286,6 +297,16 @@ private void detectMetricAnomaliesFromPeers(Map<BrokerEntity, Double> currentVal
     }
   }
 
+  private Set<BrokerEntity> detectMetricAnomaliesFromThreshold(Map<BrokerEntity, Double> currentValue, double threshold) {
+    Set<BrokerEntity> detectedMetricAnomalies = new HashSet<>();
+    for (Map.Entry<BrokerEntity, Double> entry : currentValue.entrySet()) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java:302]: The loop uses only the key, but iterates over entities. Can we either iterate over `keySet()` or update the logic to use `getValue()`, as well?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java [hunk 3]
```diff
@@ -416,6 +437,21 @@ public void configure(Map<String, ?> configs) {
       }
     }
 
+    String logFlushTimeThresholdMs = (String) originalConfig.get(SLOW_BROKER_LOG_FLUSH_TIME_THRESHOLD_MS_CONFIG);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java:440]: This function (i.e. configure) is full of repeated code fragments. Can we use util functions to eliminate code repetition and be more concise?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinderTest.java [hunk 4]
```diff
@@ -161,6 +161,36 @@ public void testNoFalsePositiveDetection() {
     assertTrue(anomalies.isEmpty());
   }
 
+  /**
+   * Test slow broker finder does not report false positive anomaly when the broker's absolute log flush time is small.
+   */
+  @Test
+  public void testNoFalsePositiveDetection2() {
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinderTest.java:168]: Nit: Enumerating functions does not give much information about the difference between `testNoFalsePositiveDetection()` and `testNoFalsePositiveDetection2()`. Can we rename them with a suffix that describes what each function are?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java (hunk 1 above)
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/SlowBrokerFinder.java:187]: I rename the method to make it more specific.
I was thinking about merge this part of logic into `getMetricAnomalies()`. The issue is the logic in `getMetricAnomalies()` is generic and this extra check is specific to log flush time metrics.

**Pull request merged**